### ISSUE TITLE
Writing Profile to Different Locations

### ISSUE THREAD

**Issue by kantai** [blockstack/blockstack-browser#1154]
A subtle result of issue #1112 is that a given identity's profile URL will be written to different locations when the bug is fixed. For example, let's say I have 3 identities each with a profile:

Before bugfix:

```
ID A: Address = m/888'/0'/0'
         URL = https://gaia.blockstack.org/hub/1DHvWDj834zPAkwMhpXdYbCYh4PomwQfzz/0/profile.json
ID B: Address = m/888'/0'/1'
         URL = https://gaia.blockstack.org/hub/1DHvWDj834zPAkwMhpXdYbCYh4PomwQfzz/1/profile.json
ID C: Address = m/888'/0'/3'
         URL = https://gaia.blockstack.org/hub/1DHvWDj834zPAkwMhpXdYbCYh4PomwQfzz/2/profile.json
```

After bugfix:
```
ID A: Address = m/888'/0'/0'
         URL = https://gaia.blockstack.org/hub/1DHvWDj834zPAkwMhpXdYbCYh4PomwQfzz/0/profile.json
ID B: Address = m/888'/0'/1'
         URL = https://gaia.blockstack.org/hub/1DHvWDj834zPAkwMhpXdYbCYh4PomwQfzz/1/profile.json
ID D: Address = m/888'/0'/2'
         URL = https://gaia.blockstack.org/hub/1DHvWDj834zPAkwMhpXdYbCYh4PomwQfzz/2/profile.json
ID C: Address = m/888'/0'/3'
         URL = https://gaia.blockstack.org/hub/1DHvWDj834zPAkwMhpXdYbCYh4PomwQfzz/3/profile.json
```

There's _two_ issues above:
1. ID C's profile URL changed. If there's a name associated with the ID, that means that we'll need a zonefile update. Even if there's not, the user will "lose" their profile data.
2. ID D's profile URL is the same as ID C's old one! So even if we solve the previous problem by allowing ID C to write to the old location, we have an issue where two profiles want to be at the same location.

_Proposed solution_
1. Browser should, in the case of profiles which _don't_ have names and zonefiles, always write to the "correct URL"
2. In the case of profiles which _do_ have names and zonefiles, _try_ to write to the URL in the zonefile. If it cannot, raise an error.
3. In the case of profiles which _don't_ have names and zonefiles, read profiles from the first profile validly signed by the expected address in the set of "once correct" profile locations -- for ID C above, that would be the list `['https://.../3/profile.json', 'https://.../2/profile.json']`
4. In the case of profiles which _do_ have names and zonefiles, browser should always read from the URL in the zonefile.
5. We should changed the profile URL to `https://gaia.blockstack.org/hub/<id-address>/profile.json` -- items 1-4 will allow us to make this change somewhat seamlessly.

This proposed solution allows us to remain backwards compatible for people who have names and zonefiles. It will migrate users who don't have zonefiles automatically (though, in some corner cases, you can end up with a profile which has been written by another identity address which will cause the migration to fail). The only case it _doesn't_ solve for is users who _already have_ zonefiles pointing to the same profile --- this case can't be solved without a zonefile update anyways. Also, it avoids new profiles from conflicting with old ones with zonefiles (which cannot be automatically migrated) by doing away with the index based URL generation in the first place.

Anyways, I don't think we should fix #1112 before we have some scheme in place for automatic migration.

Tagging @larrysalibra @yknl and @jackzampolin for some thoughts on tackling this issue.

**Comment by larrysalibra** [comment]
Thanks for looking into this @kantai - agree with your plan of attack.

> We should changed the profile URL to https://gaia.blockstack.org/hub/<id-address>/profile.json -- items 1-4 will allow us to make this change somewhat seamlessly.

Writing profiles with indexes was always intended as a temporary measure. We wanted to use this schema originally but didn't do this originally was because of the limitation of blockstack core's storage implementation which required 1 storage destination per running core instance and our time constraint limitations.

> Anyways, I don't think we should fix #1112 before we have some scheme in place for automatic migration.
 
Agree.

**Comment by kantai** [comment]
This plan was implemented and shipped.
